Span processor clarifications (#1135)
open-telemetry/opentelemetry-specification: specification/trace/sdk.md

[`IsRecording`](api.md#isrecording) is true.

Built-in span processors are responsible for batching and conversion of spans to

> exportable representation and passing batches to exporters or alternatives to

Reviewer: Not sure if a GA spec should point to an experimental spec, thoughts?

Author: Would it be true then to say "and passing batches to exporters or other span processors," or are there still other alternatives?

Author: Also, to @reyang's point, I am not even sure referring to zPages specifically makes sense here. My only goal is to have a clear definition of span processors in this sentence. Reading this section, I was confused by the fact that span processors were defined as "responsible for... passing batches to exporters" but then a few lines down, exporters are described as optional.

Reviewer 2: I don't see a problem referring to experimental feature as it makes the point that this is an extensibility point in SDK that can be useful for purposes including experimentation with the new features.

exporters such as [z-pages](../../experimental/trace/zpages.md).

Span processors can be registered directly on SDK `TracerProvider` and they are
invoked in the same order as they were registered.

> Each processor registered on `TracerProvider` is a start of a pipeline that consists
> of one or more span processors and, optionally, one or more exporters. The SDK MUST
> allow ending each pipeline with an individual exporter.
> 

Reviewer 3: I think we should remove this paragraph. Or do you know what is meant by it? AFAIK the TracerProvider must allow the registration of multiple SpanProcessors, but each built-in span processor only supports a single exporter.

```suggestion
```

Reviewer 4: I think we definitely need some better description of the possible compositions of `SpanProcessors` and `SpanExporters`. E.g. we MUST indicate somewhere that there should be a `SpanProcessor` before each `SpanExporter` (purely from methods signatures point of view). We also MUST stress and clearly describe that multiple registered `SpanProcessor` are "siblings" and do NOT call each other. All this has already created some confusion in the past.

Reviewer 5: I support removing this paragraph. We don't need to specify that pipelines end in exporters, just that the two builtin span processors end in exporters. The metric SDK hasn't thus far mentioned support for multiple processors, and why should we? It's possible to define a multi-processor that calls a series of individual processors. Should the default trace SDK include a builtin multi-processor? I suppose they should if it's common to implement multiple trace exporters with different processor settings, but that does not sound common. Wouldn't you prefer to have a single processor call multiple exporters? Why do we need to specify these things?

> multiple registered SpanProcessor are "siblings" and do NOT call each other

If you only specify SDK support for a single processor, this problem vanishes. Potentially you can also specify a multi-processor and/or a multi-exporter, but are you sure these are commonly needed?

Author: For what it's worth, as someone relatively new to OpenTelemetry concepts, it was helpful to me to have a textual description the span processor pipeline in addition to the diagram that also exists in this section. There may be some additional nuance that needs to be added, but having some form of this paragraph seems like a good idea to me.

> The SDK MUST allow users to implement and configure custom processors and decorate
> built-in processors for advanced scenarios such as tagging or filtering.
> 

Reviewer 3: I think we should remove this paragraph too. Unless someone can tell what "tagging" means here or how the SDK can allow "filtering". CC @lmolkova 

```suggestion
```

Author: Personally, I don't feel qualified to recommend whether this paragraph should be removed. I wasn't trying to clarify any concepts here, just cleaning up. Hopefully others can weigh in.

Reviewer 4: The paragraph itself is needed, at least the part about adding custom processors. Examples "tagging or filtering" are indeed misleading and may be removed.

[`IsRecording`](api.md#isrecording) is true.

Built-in span processors are responsible for batching and conversion of spans to

> exportable representation and passing batches to exporters or alternatives to

Reviewer 6: I think we are mixing 2 different things in this sentence:
1. What is a SpanProcessor, and how can this be used to interact with the SDK.
  - The SpanProcessor in general does not need to batch or talk to the exporter, it is a simple "hook" that allows custom processors to interact with the Span lifetime events (on start and on end). For example somebody can implement a processor that tracks all the active requests (one of the functionality that we want to offer in the zPages). Others can implement different other functionality.
2. What are the "built-in" SpanProcessors.
  - We will provide 2 flavors of SpanProcessors (Simple/Batch) that are able to batch (or not in case of simple), transform the Spans in an "exportable" format, and call the "exporter".

The standard OpenTelemetry SDK MUST implement both simple and batch processors,
as described below. Other common processing scenarios should be first considered
for implementation out-of-process in an [OpenTelemetry Collector](../overview.md#collector).

#### Simple processor

This is an implementation of `SpanProcessor` which passes finished spans
and the export-friendly span data representation to the configured
`SpanExporter` as soon as they are finished.

> Typically, the simple processor will be most suitable for use in testing. Exceptions to this

Reviewer 2: I don't want to position these non-testing scenarios as exceptions. These should be used with caution, but they are valid use cases for production use

> Typically, the simple processor will be most suitable for use in testing; it should be used with

Reviewer 7: Not sure this should be mentioned here.
We have production systems running using SimpleProcessors. In those systems, the OpenTelemetry SDK uses SimpleProcessor and exports the span as they arrive, without any batching. 
The export may be done to some agent, which in turn may do batching.
Or Export could be done to the actual backend, to enable scenarios like "Live telemetry".  Eg in Azure: https://docs.microsoft.com/azure/azure-monitor/app/live-stream

Reviewer 2: Added couple suggestions. @anuraaga, @cijothomas are these suggestions looks OK for you?

Author: Thank you for adding more suggestions, @SergeyKanzhelev. I have incorporated the first (though happy to revert or refine if there are further comments).

The second is still open and I'm hoping @Oberon00 or others can help clarify a more accurate version.

Thank you all!

> Typically, the simple processor will be most suitable for use in testing; it should be used with
> caution in production. It may be appropriate for production use in scenarios where creating

Reviewer 2: ```suggestion
Typically, the simple processor will be most suitable for use in testing; it should be used with
The simple processor must not be used lightly in production. It runs on application thread and may
degrade performance. It may be appropriate for production use in scenarios where creating
```

multiple threads is not desirable (since batching span processors will typically be threaded
in order to pass spans on to exporters in the background) as well as scenarios where different

> custom attributes should be added to individual spans based on code scopes.
> 

Reviewer 2: ```suggestion
custom attributes should be added to individual spans based on code scopes. Simple processors
might also be used for scenarios where the callback needs to be called before sampling. For
example, z-pages can be implemented this way.

```

Reviewer 3: This is wrong though. The simple processor is also not invoked for DROPped spans.

Reviewer 2: hm, I'm referring to the sampled flag as oppose to the recorded flag. Recorded, but sampled out spans should not make it to the batch processor, correct? Recorded will be processed by simple processors. dropped will not reach either

The standard OpenTelemetry SDK MUST implement both simple and batch processors,
as described below. Other common processing scenarios should be first considered
for implementation out-of-process in an [OpenTelemetry Collector](../overview.md#collector).

#### Simple processor

This is an implementation of `SpanProcessor` which passes finished spans
and the export-friendly span data representation to the configured
`SpanExporter` as soon as they are finished.

> Typically, the simple processor will be most suitable for use in testing and/or

Reviewer 2: it may be used in production for adding attributes like here: https://medium.com/opentelemetry/opentelemetry-beyond-getting-started-5ac43cd0fe26#ffbb:~:text=Custom%20attributes%20in%20code%20scopes

Author: Thanks for your valuable feedback on this section as well as the batch processor one, @Oberon00. Do you think more should be added to clarify that while these statements may "typically" be true, it also depends on language implementation? I'm not sure how much detail to get into here.

in scenarios where creating multiple threads is not desirable, since batching
span processors will typically be threaded in order to pass spans on to exporters
in the background.

**Configurable parameters:**

* `exporter` - the exporter where the spans are pushed.

#### Batching processor

> This is an implementation of `SpanProcessor` which creates batches of finished
> spans and passes the export-friendly span data representations to the
> configured `SpanExporter`.

Reviewer 3: ```suggestion
This is an implementation of `SpanProcessor` which collects finished spans into batches
and passes the export-friendly span data representations to the associated `SpanExporter`.
```

Typically, the batching processor will be more suitable for production environments

> than the simple processor.

Reviewer 3: Depends on the language, e.g. in Java exporters do batching themselves sometimes. But typically, I think that BatchSpanProcessor should be the expected scenario, yes.

Reviewer 2: I like how it's written. Wdyt @Oberon00?

Reviewer 3: I think the PR's wording is fine here 👍

**Configurable parameters:**

* `exporter` - the exporter where the spans are pushed.
* `maxQueueSize` - the maximum queue size. After this limit is reached, spans are

>   dropped. The default value is `2048`.

Reviewer 3: This is now misleading because the spans cannot really be dropped (as in sampled out) because the SpanProcessor has no back-channel to the TracerProvider. Would be nice if this was possible, but for now I'd say

```suggestion
  skipped by this BatchSpanProcessor. The default value is `2048`.
```
